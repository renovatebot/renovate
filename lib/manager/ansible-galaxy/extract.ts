import * as datasourceGalaxy from '../../datasource/galaxy';
import * as datasourceGitTags from '../../datasource/git-tags';
import { logger } from '../../logger';
import { SkipReason } from '../../types';
import type { PackageDependency, PackageFile } from '../types';

const galaxyUrlRegex = /((git\+)?((git|ssh|http(s)?):\/\/)?(.*@)?([\w.-]+)((:\d+)?\/|:))([\w./-]+)(\.git)?/;
const galaxyRoleRegex = /[\w-]+\.[\w-]+/;

function interpretLine(
  lineMatch: RegExpMatchArray,
  lineNumber: number,
  dependency: PackageDependency
): PackageDependency {
  const localDependency: PackageDependency = dependency;
  const key = lineMatch[2];
  const value = lineMatch[3].replace(/["']/g, '');
  switch (key) {
    case 'name': {
      localDependency.managerData.name = value;
      break;
    }
    case 'version': {
      localDependency.managerData.version = value;
      localDependency.currentValue = value;
      localDependency.managerData.lineNumber = lineNumber;
      break;
    }
    case 'scm': {
      localDependency.managerData.scm = value;
      break;
    }
    case 'src': {
      localDependency.managerData.src = value;
      break;
    }
    default: {
      return null;
    }
  }
  return localDependency;
}

function finalize(dependency: PackageDependency): boolean {
  const dep = dependency;
  if (dependency.managerData.version === null) {
    dep.skipReason = SkipReason.NoVersion;
    return false;
  }

  const source: string = dep.managerData.src;
  const sourceMatch: RegExpMatchArray = galaxyUrlRegex.exec(source);
  if (sourceMatch) {
    dep.datasource = datasourceGitTags.id;
    // remove ending `.git` from URLs
    dep.depName = sourceMatch[10].replace(/\.git$/, '');
    // remove leading `git+` from URLs like `git+https://...`
    dep.lookupName = source.replace(/git\+/, '');
  } else if (galaxyRoleRegex.exec(source)) {
    dep.datasource = datasourceGalaxy.id;
    dep.depName = dep.managerData.src;
    dep.lookupName = dep.managerData.src;
  } else if (galaxyRoleRegex.exec(dep.managerData.name)) {
    dep.datasource = datasourceGalaxy.id;
    dep.depName = dep.managerData.name;
    dep.lookupName = dep.managerData.name;
  } else {
    dep.skipReason = SkipReason.NoSourceMatch;
    return false;
  }
  if (dep.managerData.name !== null) {
    dep.depName = dep.managerData.name;
  }

  return true;
}

export default function extractPackageFile(
  content: string
): PackageFile | null {
  logger.trace('ansible-galaxy.extractPackageFile()');
  const newBlockRegEx = new RegExp(/^\s*-\s*((\w+):\s*(.*))$/);
  const blockLineRegEx = new RegExp(/^\s*((\w+):\s*(.*))$/);
  const deps: PackageDependency[] = [];
  const lines = content.split('\n');
  try {
    for (let lineNumber = 0; lineNumber < lines.length; lineNumber += 1) {
      let lineMatch = newBlockRegEx.exec(lines[lineNumber]);
      if (lineMatch) {
        const dep: PackageDependency = {
          managerData: {
            name: null,
            version: null,
            scm: null,
            src: null,
          },
        };
        do {
          const localdep = interpretLine(lineMatch, lineNumber, dep);
          if (localdep == null) {
            break;
          }
          const line = lines[lineNumber + 1];

          if (!line) {
            break;
          }
          lineMatch = blockLineRegEx.exec(line);
          if (lineMatch) {
            lineNumber += 1;
          }
        } while (lineMatch);
        if (finalize(dep)) {
          delete dep.managerData;
          deps.push(dep);
        }
      }
    }

    if (!deps.length) {
      return null;
    }
    return { deps };
  } catch (err) /* istanbul ignore next */ {
    logger.debug({ err }, 'Error extracting ansible-galaxy deps');
    return null;
  }
}
