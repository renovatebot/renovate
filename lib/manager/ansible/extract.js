const { getDep } = require('../dockerfile/extract');

module.exports = {
  extractPackageFile,
};

function extractPackageFile(content) {
  logger.debug('ansible.extractPackageFile()');
  let deps = [];
  let lineNumber = 0;
  for (const line of content.split('\n')) {
    const match = line.match(/^\s*image:\s*'?"?([^\s'"]+)'?"?\s*$/);
    if (match) {
      const currentFrom = match[1];
      const dep = getDep(currentFrom);
      logger.debug(
        {
          depName: dep.depName,
          currentValue: dep.currentValue,
          currentDigest: dep.currentDigest,
        },
        'Docker image inside ansible'
      );
      dep.lineNumber = lineNumber;
      dep.versionScheme = 'docker';
      deps.push(dep);
    }
    lineNumber += 1;
  }
  deps = deps.filter(
    dep => !(dep.currentValue && dep.currentValue.includes('${'))
  );
  if (!deps.length) {
    return null;
  }
  return { deps };
}
