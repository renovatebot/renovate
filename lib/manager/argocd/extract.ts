import { loadAll } from 'js-yaml';
import * as gitTags from '../../datasource/git-tags';
import { HelmDatasource } from '../../datasource/helm';
import type { ExtractConfig, PackageDependency, PackageFile } from '../types';
import type { ApplicationDefinition } from './types';
import { fileTestRegex } from './util';

function createDependency(
  definition: ApplicationDefinition
): PackageDependency {
  const source = definition.spec?.source;

  if (source == null) {
    return null;
  }

  // a chart variable is defined this is helm declaration
  if (source.chart) {
    return {
      depName: source.chart,
      registryUrls: [source.repoURL],
      currentValue: source.targetRevision,
      datasource: HelmDatasource.id,
    };
  }
  return {
    depName: source.repoURL,
    currentValue: source.targetRevision,
    datasource: gitTags.id,
  };
}

export function extractPackageFile(
  content: string,
  fileName: string,
  config?: ExtractConfig
): PackageFile | null {
  // check for argo reference. API version for the kind attribute is used
  if (fileTestRegex.test(content) === false) {
    return null;
  }

  const definitions: ApplicationDefinition[] = loadAll(content);

  const deps = definitions
    .map((definition) => createDependency(definition))
    .filter(Boolean);

  return deps.length ? { deps } : null;
}
