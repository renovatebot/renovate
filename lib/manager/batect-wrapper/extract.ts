import { id as githubReleaseDatasource } from '../../datasource/github-releases';
import { logger } from '../../logger';
import { id as semverVersioning } from '../../versioning/semver';
import type { PackageDependency, PackageFile } from '../types';

const VERSION_REGEX = /^\s+VERSION="(.*)"$/m;

export function extractPackageFile(fileContent: string): PackageFile | null {
  logger.trace('batect.extractPackageFile()');
  const match = VERSION_REGEX.exec(fileContent);

  if (match === null) {
    return null;
  }

  const dependency: PackageDependency = {
    depName: 'batect/batect',
    commitMessageTopic: 'Batect',
    currentValue: match[1],
    datasource: githubReleaseDatasource,
    versioning: semverVersioning,
  };

  logger.trace(dependency, 'Found Batect wrapper version');

  return { deps: [dependency] };
}
