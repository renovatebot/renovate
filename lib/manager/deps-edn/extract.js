const { DEFAULT_MAVEN_REPO } = require('../maven/extract');
const { expandDepName, DEFAULT_CLOJARS_REPO } = require('../leiningen/extract');

function extractPackageFile(content) {
  const deps = [];

  const regex = /([^{\s,]*)[\s,]*{[\s,]*:mvn\/version[\s,]+"([^"]+)"[\s,]*}/;
  let rest = content;
  let match = rest.match(regex);
  let offset = 0;
  while (match) {
    const [wholeSubstr, depName, currentValue] = match;
    const fileReplacePosition =
      offset + match.index + wholeSubstr.indexOf(currentValue);

    offset += match.index + wholeSubstr.length;
    rest = content.slice(offset);
    match = rest.match(regex);

    deps.push({
      datasource: 'maven',
      depName: expandDepName(depName),
      currentValue,
      fileReplacePosition,
      registryUrls: [DEFAULT_CLOJARS_REPO, DEFAULT_MAVEN_REPO],
    });
  }

  return { deps };
}

module.exports = {
  extractPackageFile,
};
