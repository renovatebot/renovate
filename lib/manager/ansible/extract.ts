import { logger } from '../../logger';
import * as dockerVersioning from '../../versioning/docker';
import { PackageDependency, PackageFile } from '../common';
import { getDep } from '../dockerfile/extract';

export default function extractPackageFile(
  content: string
): PackageFile | null {
  logger.trace('ansible.extractPackageFile()');
  let deps: PackageDependency[] = [];
  const re = /^\s*image:\s*'?"?([^\s'"]+)'?"?\s*$/;
  for (const line of content.split('\n')) {
    const match = re.exec(line);
    if (match) {
      const currentFrom = match[1];
      const dep = getDep(currentFrom);
      logger.debug(
        {
          depName: dep.depName,
          currentValue: dep.currentValue,
          currentDigest: dep.currentDigest,
        },
        'Docker image inside ansible'
      );
      dep.versioning = dockerVersioning.id;
      deps.push(dep);
    }
  }
  deps = deps.filter((dep) => !dep.currentValue?.includes('${'));
  if (!deps.length) {
    return null;
  }
  return { deps };
}
