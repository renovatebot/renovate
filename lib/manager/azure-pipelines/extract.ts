import { safeLoad } from 'js-yaml';
import * as datasourceGitTags from '../../datasource/git-tags';
import { logger } from '../../logger';
import { PackageDependency, PackageFile } from '../common';
import { getDep } from '../dockerfile/extract';

interface Container {
  image: string;
}

interface Repository {
  type: 'git' | 'github' | 'bitbucket';
  name: string;
  ref: string;
}

interface Resources {
  repositories: Repository[];
  containers: Container[];
}

interface AzurePipelines {
  resources: Resources;
}

export function extractRepository(
  repository: Repository
): PackageDependency | null {
  if (repository.type !== 'github') {
    return null;
  }

  if (!repository.ref?.startsWith('refs/tags/')) {
    return null;
  }

  return {
    autoReplaceStringTemplate: 'refs/tags/{{newValue}}',
    currentValue: repository.ref.replace('refs/tags/', ''),
    datasource: datasourceGitTags.id,
    depName: repository.name,
    depType: 'gitTags',
    lookupName: `https://github.com/${repository.name}.git`,
    replaceString: repository.ref,
  };
}

export function extractContainer(
  container: Container
): PackageDependency | null {
  if (!container.image) {
    return null;
  }

  const dep = getDep(container.image);
  logger.debug(
    {
      depName: dep.depName,
      currentValue: dep.currentValue,
      currentDigest: dep.currentDigest,
    },
    'Azure pipelines docker image'
  );
  dep.depType = 'docker';

  return dep;
}

export function parseAzurePipelines(content: string): AzurePipelines | null {
  let pkg = null;
  try {
    pkg = safeLoad(content);
  } catch (err) /* istanbul ignore next */ {
    logger.error({ err }, 'Error parsing azure-pipelines content');
    return null;
  }

  if (!pkg || !pkg.resources) {
    return null;
  }

  pkg.resources.containers = pkg.resources.containers || [];
  pkg.resources.repositories = pkg.resources.repositories || [];

  return pkg;
}

export function extractPackageFile(content: string): PackageFile | null {
  logger.trace('azurePipelines.extractPackageFile()');
  const deps: PackageDependency[] = [];

  const pkg = parseAzurePipelines(content);
  if (!pkg) {
    return null;
  }

  // grab the repositories tags
  for (const repository of pkg.resources.repositories) {
    const dep = extractRepository(repository);
    if (dep) {
      deps.push(dep);
    }
  }

  // grab the containers tags
  for (const container of pkg.resources.containers) {
    const dep = extractContainer(container);
    if (dep) {
      deps.push(dep);
    }
  }

  if (!deps.length) {
    return null;
  }
  return { deps };
}
