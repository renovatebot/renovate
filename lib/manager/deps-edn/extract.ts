import { DEFAULT_MAVEN_REPO } from '../maven/extract';
import { expandDepName, DEFAULT_CLOJARS_REPO } from '../leiningen/extract';
import { PackageFile, PackageDependency } from '../common';
import { DATASOURCE_MAVEN } from '../../constants/data-binary-source';

export function extractPackageFile(content: string): PackageFile {
  const deps: PackageDependency[] = [];

  const regex = /([^{\s,]*)[\s,]*{[\s,]*:mvn\/version[\s,]+"([^"]+)"[\s,]*}/;
  let rest = content;
  let match = regex.exec(rest);
  let offset = 0;
  while (match) {
    const [wholeSubstr, depName, currentValue] = match;
    const fileReplacePosition =
      offset + match.index + wholeSubstr.indexOf(currentValue);

    offset += match.index + wholeSubstr.length;
    rest = content.slice(offset);
    match = regex.exec(rest);

    deps.push({
      datasource: DATASOURCE_MAVEN,
      depName: expandDepName(depName),
      currentValue,
      fileReplacePosition,
      registryUrls: [DEFAULT_CLOJARS_REPO, DEFAULT_MAVEN_REPO],
    });
  }

  return { deps };
}
