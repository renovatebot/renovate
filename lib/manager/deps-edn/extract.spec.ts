/* eslint-disable no-template-curly-in-string */
import { readFileSync } from 'fs';
import { resolve } from 'upath';
import { extractPackageFile } from './extract';

const depsEdn = readFileSync(
  resolve(__dirname, `./__fixtures__/deps.edn`),
  'utf8'
);

describe('manager/deps-edn/extract', () => {
  it('extractPackageFile', () => {
    expect(extractPackageFile(depsEdn)).toMatchSnapshot();
  });
});
