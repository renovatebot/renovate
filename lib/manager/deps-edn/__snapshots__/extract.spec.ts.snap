// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`manager/deps-edn/extract extractPackageFile 1`] = `
Object {
  "deps": Array [
    Object {
      "currentValue": "0.1.2",
      "datasource": "maven",
      "depName": "persistent-sorted-set:persistent-sorted-set",
      "fileReplacePosition": 53,
      "registryUrls": Array [
        "https://clojars.org/repo/",
        "https://repo.maven.apache.org/maven2",
      ],
    },
    Object {
      "currentValue": "1.9.0",
      "datasource": "maven",
      "depName": "org.clojure:clojure",
      "fileReplacePosition": 147,
      "registryUrls": Array [
        "https://clojars.org/repo/",
        "https://repo.maven.apache.org/maven2",
      ],
    },
    Object {
      "currentValue": "1.10.0",
      "datasource": "maven",
      "depName": "org.clojure:clojure",
      "fileReplacePosition": 241,
      "registryUrls": Array [
        "https://clojars.org/repo/",
        "https://repo.maven.apache.org/maven2",
      ],
    },
    Object {
      "currentValue": "1.10.520",
      "datasource": "maven",
      "depName": "org.clojure:clojurescript",
      "fileReplacePosition": 389,
      "registryUrls": Array [
        "https://clojars.org/repo/",
        "https://repo.maven.apache.org/maven2",
      ],
    },
    Object {
      "currentValue": "0.2.11",
      "datasource": "maven",
      "depName": "org.clojure:tools.namespace",
      "fileReplacePosition": 451,
      "registryUrls": Array [
        "https://clojars.org/repo/",
        "https://repo.maven.apache.org/maven2",
      ],
    },
    Object {
      "currentValue": "1.10.520",
      "datasource": "maven",
      "depName": "org.clojure:clojurescript",
      "fileReplacePosition": 584,
      "registryUrls": Array [
        "https://clojars.org/repo/",
        "https://repo.maven.apache.org/maven2",
      ],
    },
    Object {
      "currentValue": "0.0-389",
      "datasource": "maven",
      "depName": "lambdaisland:kaocha",
      "fileReplacePosition": 644,
      "registryUrls": Array [
        "https://clojars.org/repo/",
        "https://repo.maven.apache.org/maven2",
      ],
    },
    Object {
      "currentValue": "0.0-21",
      "datasource": "maven",
      "depName": "lambdaisland:kaocha-cljs",
      "fileReplacePosition": 703,
      "registryUrls": Array [
        "https://clojars.org/repo/",
        "https://repo.maven.apache.org/maven2",
      ],
    },
    Object {
      "currentValue": "0.21.1",
      "datasource": "maven",
      "depName": "cider:cider-nrepl",
      "fileReplacePosition": 810,
      "registryUrls": Array [
        "https://clojars.org/repo/",
        "https://repo.maven.apache.org/maven2",
      ],
    },
    Object {
      "currentValue": "0.6.0",
      "datasource": "maven",
      "depName": "nrepl:nrepl",
      "fileReplacePosition": 870,
      "registryUrls": Array [
        "https://clojars.org/repo/",
        "https://repo.maven.apache.org/maven2",
      ],
    },
    Object {
      "currentValue": "0.2.11",
      "datasource": "maven",
      "depName": "org.clojure:tools.namespace",
      "fileReplacePosition": 929,
      "registryUrls": Array [
        "https://clojars.org/repo/",
        "https://repo.maven.apache.org/maven2",
      ],
    },
    Object {
      "currentValue": "0.9.5703",
      "datasource": "maven",
      "depName": "com.datomic:datomic-free",
      "fileReplacePosition": 1141,
      "registryUrls": Array [
        "https://clojars.org/repo/",
        "https://repo.maven.apache.org/maven2",
      ],
    },
  ],
}
`;
